src/main/scala/com/microsoft/hyperspace/index/dataskipping/util/SortedArrayLowerBound.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (2021) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.dataskipping.util
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{BinaryExpression, Expression}
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, CodeGenerator, ExprCode, FalseLiteral}
+import org.apache.spark.sql.catalyst.expressions.codegen.Block._
+import org.apache.spark.sql.catalyst.util.{ArrayData, TypeUtils}
+import org.apache.spark.sql.types.IntegerType
+
+/**
+ * Returns the index to the first element in the left array which is not less
+ * than (greater than or equal to) the right value, or null if there is no such
+ * element.
+ *
+ * The array must not be null.
+ * Elements in the array must be in ascending order.
+ * The array must not contain null elements.
+ * The array must not contain duplicate elements.
+ * The value must not be null.
+ */
+case class SortedArrayLowerBound(left: Expression, right: Expression) extends BinaryExpression {
```

Comment:
Could you add `private[hyperspace]` to this class as it has the unchecked limitations

Reply:
Done